### DISCUSSION TITLE
File reader and file writer with FTP server

### DISCUSSION THREAD

**Discussion by michelangelo404** [nextgenhealthcare/connect#5361] [Q&A]
Hi, I am a newbie and I am using mirth for a client who needs to check how many lines have been inserted in a text file located in my ftp server, what I am trying to do is read the contents of the file, and then send the number of lines to another mirth channel, which will have the task of writing the number of lines to another log file using the file writer connector.
what is the best way to do this? should I use javascript?
Thank you in advance.

**Comment by jonbartels** [comment]
Do you just want a count of the lines or do you want the actual contents of each line?

Are you using FTP or are you using SFTP? (It matters since SFTP can sometimes give you more shell commands since its SSH based. )

How large is the file?

**Reply by michelangelo404** [reply]
I want to count the number of lines, i'm using FTP and the maximum file size is 3 mb.

**Reply by jonbartels** [reply]
I think for FTP you would have to get the entire file. 

The way to do the FTP read in MC is a simple file reader.
Get the entire file then use a transformer step to count the lines (split on line breaks and count it up)
Persist the line count to a globalChannelMap variable
Write the count as needed to downstream systems
On the next poll take the difference of the current count from your globalChannelMap variable

So the first poll might be 100 lines, then the next poll counts 200 lines (so net-new of 100) and so on.

The ideal case would be something like SSH then run `wc -l` on the REMOTE server so very little data is transferred. I don't think you can do that with FTP though.

**Reply by pacmano1** [reply]
The OP's original requirement is arguably a very poor design since presumably the file will grow forever and you must transport the entire file to count the total number of lines to determine the difference from the prior poll.

The more typical way would be to create a new file for new rows irrespective of your need to create some large file (which could still be some other process/file).    Or some other timed chunking method.

**Comment by VivekBIData** [comment]
Hi @jonbartels/ @pacmano1,

I have been trying to connect with sftp using password creds but I am unable to connect it.
When I am trying to test the connection, it is giving me below error as "Auth Fail".

Could you help here? Is the open source version available to connect with GCS?

![image](https://github.com/nextgenhealthcare/connect/assets/99861749/434ac02e-679e-4f32-b94a-53bcef21c7f7)

**Reply by jonbartels** [reply]
https://github.com/nextgenhealthcare/connect/issues/5793 is a better place to get an answer to your question.
